Clean-up message_handler of parse_options_baset

romainbrenguier opened the pull request:
~Based on https://github.com/diffblue/cbmc/pull/4520~

The message handler field was owned by the classes inheriting from
parse_options_baset which lead to strange patterns where the handler
reference was initialized using a reference to a field which was not
constructed yet.
This is cleaner and avoid some reference duplication.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [na] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
Make parse_options_baset the owner of message_handler

The message handler field was owned by the classes inheriting from
parse_options_baset which lead to strange patterns where the handler
reference was initialized using a reference to a field which was not
constructed yet.
This is cleaner and avoid some reference duplications.

romainbrenguier committed:
Remove comment made unecessary

Now the log field gets initialized before parse_options_baset is
constructed so there is no danger as described in the comment.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 2db3a82183b5daa5ea153822b981dde8c1414f8e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108107188
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: a76ee1fd4af4ec1b50f7a6a272d3b6894840962b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108108467
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 0e56317b66922663bb6ca975461dc263f4cb477f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108123437
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

tautschnig commented on src/util/parse_options.h:
```diff
@@ -35,6 +37,7 @@ class parse_options_baset
   virtual ~parse_options_baset() { }
 
 protected:
+  std::unique_ptr<ui_message_handlert> message_handler;
```
Why can't this just be a stack-allocated member?

tautschnig commented on src/util/parse_options.cpp:
```diff
@@ -28,11 +28,12 @@ parse_options_baset::parse_options_baset(
   std::string optstring=std::string("?h(help)")+_optstring;
   parse_result=cmdline.parse(argc, argv, optstring.c_str());
+  message_handler = util_make_unique<ui_message_handlert>(cmdline, program);
+  log = messaget{*message_handler};
```
I really think that `message_handler` should be a full member (see also comment below) and the constructor should do `: message_handler(cmdline, program), log(message_handler)`

romainbrenguier replied:
I've done the change now. The reason I didn't do it originally was that it required a bit of reordering because the command line needs to be parsed before the object is constructed, but the change is actually not too bad.

tautschnig reviewed (commented):
The second commit alone is worth merging this as soon as possible, but I really think we should not need to use `std::unique_ptr` here.

tautschnig commented on jbmc/src/janalyzer/janalyzer_parse_options.h:
```diff
@@ -174,7 +173,7 @@ class janalyzer_parse_optionst : public parse_options_baset
   ui_message_handlert::uit get_ui()
   {
-    return ui_message_handler.get_ui();
+    return message_handler.get_ui();
```
Is this method (`janalyzer_parse_optionst::get_ui`) actually used?

romainbrenguier replied:
I think it's not useful anyway...

tautschnig replied:
Maybe just add a commit that removes those?

tautschnig commented on jbmc/src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -564,7 +566,7 @@ int jbmc_parse_optionst::doit()
     {
       show_properties(
-        goto_model, log.get_message_handler(), ui_message_handler.get_ui());
+        goto_model, log.get_message_handler(), message_handler.get_ui());
```
🍌 Are `log.get_message_handler()` and `message_handler` actually different?

romainbrenguier replied:
I hope not. But I will check.

peterschrammel replied:
Message_handler and ui shouldn't be passed separately.

tautschnig commented on jbmc/src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -800,7 +802,7 @@ int jbmc_parse_optionst::get_goto_program(
         *goto_model,
         log.get_message_handler(),
-        ui_message_handler.get_ui(),
+        message_handler.get_ui(),
```
🍌 Again, are `log.get_message_handler()` and `message_handler` different? It looks quite confusing.

tautschnig commented on jbmc/src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -943,7 +944,7 @@ bool jbmc_parse_optionst::show_loaded_functions(
       ns,
       log.get_message_handler(),
-      ui_message_handler.get_ui(),
+      message_handler.get_ui(),
```
🍌

tautschnig commented on jbmc/src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -930,8 +931,8 @@ bool jbmc_parse_optionst::show_loaded_functions(
-      ui_message_handler,
-      ui_message_handler.get_ui(),
+      message_handler,
+      message_handler.get_ui(),
```
Relates to 🍌

tautschnig commented on jbmc/src/jdiff/jdiff_parse_options.cpp:
```diff
@@ -226,12 +229,12 @@ int jdiff_parse_optionst::doit()
       goto_model1,
       log.get_message_handler(),
-      ui_message_handler.get_ui(),
+      message_handler.get_ui(),
```
🍌

tautschnig commented on jbmc/src/jdiff/jdiff_parse_options.cpp:
```diff
@@ -226,12 +229,12 @@ int jdiff_parse_optionst::doit()
       goto_model2,
       log.get_message_handler(),
-      ui_message_handler.get_ui(),
+      message_handler.get_ui(),
```
🍌

tautschnig commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -545,7 +547,7 @@ int cbmc_parse_optionst::doit()
 
   int get_goto_program_ret =
-    get_goto_program(goto_model, options, cmdline, log, ui_message_handler);
+    get_goto_program(goto_model, options, cmdline, log, message_handler);
```
Are we actually passing a `messaget` and a `message_handlert` here?!

romainbrenguier replied:
the function needs a `messaget` and a `ui_message_handlert`, so `log.get_message_handler()` wouldn't have the right type.

tautschnig replied:
This isn't at all a questionable piece of code you're introducing here, but I'm really wondering why we have this interface. It should be sufficient to pass a `message_handlert` or `ui_message_handlert` and then just create a `messaget` locally in that function.

peterschrammel replied:
`get_goto_program` is a non-static member, so neither log nor message_handler should be passed.

romainbrenguier replied:
`get_goto_program` is static in `cbmc_parse_optionst`. I will add a commit to construct the messaget from the message handler instead of passing both.

tautschnig commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -554,7 +556,7 @@ int cbmc_parse_optionst::doit()
   {
     show_properties(
-      goto_model, log.get_message_handler(), ui_message_handler.get_ui());
+      goto_model, log.get_message_handler(), message_handler.get_ui());
```
🍌

tautschnig commented on src/goto-analyzer/goto_analyzer_parse_options.h:
```diff
@@ -177,7 +176,7 @@ class goto_analyzer_parse_optionst: public parse_options_baset
   ui_message_handlert::uit get_ui()
   {
-    return ui_message_handler.get_ui();
+    return message_handler.get_ui();
```
Is `goto_analyzer_parse_optionst::get_ui` actually used?

tautschnig commented on src/goto-diff/goto_diff_parse_options.cpp:
```diff
@@ -264,12 +267,12 @@ int goto_diff_parse_optionst::doit()
       goto_model1,
       log.get_message_handler(),
-      ui_message_handler.get_ui(),
+      message_handler.get_ui(),
```
🍌

tautschnig commented on src/goto-diff/goto_diff_parse_options.cpp:
```diff
@@ -264,12 +267,12 @@ int goto_diff_parse_optionst::doit()
       goto_model2,
       log.get_message_handler(),
-      ui_message_handler.get_ui(),
+      message_handler.get_ui(),
```
🍌

tautschnig commented on src/goto-harness/goto_harness_main.cpp:
```diff
@@ -10,6 +10,6 @@ Author: Diffblue Ltd.
 {
-  auto parse_options = goto_harness_parse_optionst(argc, argv);
+  goto_harness_parse_optionst parse_options{argc, argv};
   return parse_options.main();
```
While at it, go one step further and do `return goto_harness_parse_optionst{argc, argv}.main();`?

tautschnig commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -617,7 +617,7 @@ int goto_instrument_parse_optionst::doit()
         goto_model,
         log.get_message_handler(),
-        ui_message_handler.get_ui(),
+        message_handler.get_ui(),
```
🍌

peterschrammel replied:
No need to pass message_handler and ui separately.

tautschnig commented on src/goto-instrument/goto_instrument_parse_options.h:
```diff
@@ -150,7 +148,7 @@ class goto_instrument_parse_optionst : public parse_options_baset
   ui_message_handlert::uit get_ui()
   {
-    return ui_message_handler.get_ui();
+    return message_handler.get_ui();
```
Is `goto_instrument_parse_optionst::get_ui` actually used?

tautschnig commented on src/util/parse_options.h:
```diff
@@ -10,10 +10,12 @@ Author: Daniel Kroening, kroening@kroening.com
 #ifndef CPROVER_UTIL_PARSE_OPTIONS_H
 #define CPROVER_UTIL_PARSE_OPTIONS_H
 
+#include <memory>
```
No longer needed

tautschnig commented on src/util/parse_options.h:
```diff
@@ -34,12 +36,15 @@ class parse_options_baset
+  bool parse_result;
+
 protected:
+  ui_message_handlert message_handler;
```
I think you could reduce the size of the diff quite a bit by just naming this `ui_message_handler`.

peterschrammel replied:
I also have a slight preference for `ui_message_handler` in order to distinguish it from a plain message_handler.

tautschnig reviewed (approved):
I think this is a nice bit of cleanup, but I suppose you didn't quite expect to be opening a can of worms. See various notes below.

romainbrenguier committed:
Remove redundant get_ui() method

ui_message_handler.get_ui() already does the same thing, and is more
explicit.

romainbrenguier committed:
Replace log.get_message_handler() calls

The calls will always return a reference to the field ui_message_handler
which can be accessed directly instead.

peterschrammel commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -616,7 +617,7 @@ int goto_instrument_parse_optionst::doit()
         goto_model,
-        log.get_message_handler(),
+        ui_message_handler,
         ui_message_handler.get_ui(),
```
No need to pass the ui separately (several occurrences).

romainbrenguier committed:
Remove unecessary log argument in cbmc_parse_optionst

The messaget can be constructed from the message_handler.

romainbrenguier committed:
Show properties take a ui_message_handlert

This is more precise on the type of message handler we are expecting and
makes the ui argument redundant.

romainbrenguier committed:
show_goto_functions takes a ui_message_handler

This removes the need for passing a seperate ui argument.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 79b4117a001b27dc29f42e8f1927f068d035d17f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108645501
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

romainbrenguier commented:
@peterschrammel I removed the unnecessary arguments

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 46ec5831dabd6a72cb35ae0be37bd4d2bebbc76c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108672915
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 262cc351dd74265dded3b6e4120962dc33e5fd76).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108712914
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.
